Logging Module Improvements V3

mluessi opened the pull request:
This is an improved version of https://github.com/TauLabs/TauLabs/pull/2031 Improvements:

- Configurable max logging frequency
- Logging profiles: The Default profile will just log a small subset of UAVOs (as we did previously) and the Custom profile will use the per UAVO logging period. This is useful for development to e.g. log a few UAVOs with high rates. If we add a nice way to the GCS to configure the rates, we can remove the profiles in the future.
- Modified the UAVO manager so that deregistering callbacks doesn't leak memory, i.e., the same memory can be reused for new callbacks

This WIP and has not yet been tested..

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/d-ronin/dronin/180)
<!-- Reviewable:end -->

droidicus commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[flight/Modules/Logging/logging.c, line 9 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/180#-K4x-sTwUTLRiC0mGEsQ:-K4x-sTwUTLRiC0mGEsR:1644196641) ([raw file](https://github.com/d-ronin/dronin/blob/a63277035b497d508630e01705242841480aec80/flight/Modules/Logging/logging.c#L9)):*
Add line:
* @author     dRonin, http://dronin.org Copyright (C) 2015
???

---

*[flight/PiOS/Common/pios_streamfs.c, line 4 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/180#-K4x09TU39lHcNdVfk4u:-K4x09TU39lHcNdVfk4v:1644196641) ([raw file](https://github.com/d-ronin/dronin/blob/a63277035b497d508630e01705242841480aec80/flight/PiOS/Common/pios_streamfs.c#L4)):*
Add line:
* @author     dRonin, http://dronin.org Copyright (C) 2015
???

---

*[flight/UAVObjects/uavobjectmanager.c, line 10 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/180#-K4x0Ci4ieSWNbmwoX9N:-K4x0Ci5fmmm-ToI_vIj:1644196641) ([raw file](https://github.com/d-ronin/dronin/blob/a63277035b497d508630e01705242841480aec80/flight/UAVObjects/uavobjectmanager.c#L10)):*
Add line:
* @author     dRonin, http://dronin.org Copyright (C) 2015
???

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/180)*
<!-- Sent from Reviewable.io -->

mlyle commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[flight/UAVObjects/uavobjectmanager.c, line 2076 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/180#-K4xRRem1VF8qxEWKFBp:-K4xRRem1VF8qxEWKFBq:108451368) ([raw file](https://github.com/d-ronin/dronin/blob/a63277035b497d508630e01705242841480aec80/flight/UAVObjects/uavobjectmanager.c#L2076)):*
This looks like a pool-type allocator.  Can we look into wrapping / using http://chibios.sourceforge.net/html/group__pools.html in ChibiOS instead?   (I think all targets doing logging will be on ChibiOS soon.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/180)*
<!-- Sent from Reviewable.io -->

mluessi commented:
@mlyle @droidicus thanks for the review. I fixed a bug an it is working now.

Module disabled: 
Heap: 47128 free CPU: 20%

Module enabled and idle:
Heap: 45424 free CPU: 20%

Logging at 25Hz, default profile (no other modules):
Heap: 45424 free CPU: 21%

Logging at 100Hz, default profile (no other modules):
Heap: 45424 free CPU: 24%

Rates higher than 100Hz don't work with the default profile. I did a test logging gyro data at 500Hz and it seemed to work. All tests were done on Brain.

mluessi commented:
PS: of course we don't see how much memory it uses.. see #183

mluessi commented:
Example how to parse log files in Python: https://gist.github.com/mluessi/83ab4b4ad869625401e7

mlyle commented:
@droidicus indicates he may be able to test soon.

tracernz commented:
Reviewed 1 of 4 files at r2.
Review status: 1 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[flight/UAVObjects/uavobjectmanager.c, line 2076 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/180#-K4xRRem1VF8qxEWKFBp:-K4zC6iS8pro5294qsye:1133254127) ([raw file](https://github.com/d-ronin/dronin/blob/a63277035b497d508630e01705242841480aec80/flight/UAVObjects/uavobjectmanager.c#L2076)):*
Yeah, the revo doesn't have logging until the Chibi PR merges anyway so will be fine.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/180)*
<!-- Sent from Reviewable.io -->

mluessi commented:
Review status: 1 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[flight/UAVObjects/uavobjectmanager.c, line 2076 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/180#-K4xRRem1VF8qxEWKFBp:-K4zCvkrKi_j_-9erq7a:-1309119292) ([raw file](https://github.com/d-ronin/dronin/blob/a63277035b497d508630e01705242841480aec80/flight/UAVObjects/uavobjectmanager.c#L2076)):*
OK, I can change it to use the ChibiOS pool allocator, it means logging won't work on freertos, but that would be OK IMO

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/180)*
<!-- Sent from Reviewable.io -->

mluessi commented:
Review status: 1 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[flight/UAVObjects/uavobjectmanager.c, line 2076 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/180#-K4xRRem1VF8qxEWKFBp:-K4zECvSAkuXRCjtL89g:-380479925) ([raw file](https://github.com/d-ronin/dronin/blob/a63277035b497d508630e01705242841480aec80/flight/UAVObjects/uavobjectmanager.c#L2076)):*
Silly me.. actually, this isn't in the logging module.  It's in the UAVO mananger,  so all targets would have to move to ChibiOS before the pool alloc can be used.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/180)*
<!-- Sent from Reviewable.io -->

mlyle commented:
Review status: 1 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[flight/UAVObjects/uavobjectmanager.c, line 2076 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/180#-K4xRRem1VF8qxEWKFBp:-K4zEezc5cMSZklAQNep:-1894571364) ([raw file](https://github.com/d-ronin/dronin/blob/a63277035b497d508630e01705242841480aec80/flight/UAVObjects/uavobjectmanager.c#L2076)):*
Ah, that's true.  How unfortunate.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/180)*
<!-- Sent from Reviewable.io -->

mlyle commented:
Reviewed 2 of 5 files at r1, 3 of 4 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/180)*
<!-- Sent from Reviewable.io -->

droidicus commented:
Logs are being saved in Windows as "dRonin-2015-12-07_21-00-46.drlog.tll"

kubark42 commented:
*[flight/UAVObjects/uavobjectmanager.c, line 2076 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/180#-K4xRRem1VF8qxEWKFBp:-K4zb74fpngwWTp_Rg3T:1983296060) ([raw file](https://github.com/d-ronin/dronin/blob/a63277035b497d508630e01705242841480aec80/flight/UAVObjects/uavobjectmanager.c#L2076)):*
Something to consider is that FreeRTOS has wider support on the more advanced microprocessors.  Now that there are $5 Raspberry Pis, I wouldn't be surprised if there's renewed interest in keeping FreeRTOS going.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/180)*
<!-- Sent from Reviewable.io -->

mlyle commented:
@kubark42 I am working on a raspberry pi2 port.  It requires multicore (not able to run on the original pi) and uses ChibiOS.

The advantage of the approach is it runs as a realtime process on Linux and then does its own task switching within that.  So it is possible to run low priority processes, use the Linux networking stack, etc, on the other 3 cores.

tracernz commented:
That's actually a good point @kubark42. Looks like no atmel support in chibi? That means we're pretty much committing to stm32 and closing the door on SAM E for example, no double precision fpu for us.

kubark42 commented:
The STM32 parts are underwhelming. It's disappointing, really. The SAM E packs a ton more in, and is the same (budgetary) price. However, at $5-$10 for the latest generation of microprocessor SOMs, some even including WiFi, I suspect we're going to see see migration toward the stuff like what @mlyle is working on. There new Raspberry Pi is cheaper than you can buy the newer STM32 chips for, even in 10k quantities. 

<shakes head>

I really don't understand how they got that so cheap.1 GHz processor. 512MB or RAM. Wow.

mlyle commented:
SDRAM cheap, SRAM expensive :)

droidicus commented:
Did some testing on this tonight, and was getting some really inconsistent results. I was only able to get one log to occur, and then subsequent logs were not being captured. I started having some computer issues shortly thereafter, but I don't think that would have impacted the testing.

If someone else can pull this PR and test that would be appreciated.

mluessi commented:
@droidicus thanks for the testing; I will investigate. What target were you using for your tests?

droidicus commented:
Sparky2

mlyle commented:
Tagging into renatus for now, but if it doesn't settle out quick will need to be deferred.

mluessi commented:
Rebased on latest next. Fixed the GCS filename extension and added systemident and systemalarms to the default profile. Did more test, works very well on Brain, was able to log accels at 200Hz with the custom profile and tested logging with default profile extensively.

One difference between brain and sparky2 is that brain uses 4kB sectors while sparky2 uses 64kB sectors, not sure if this is causing the problems @droidicus observed. Can you do another test with sparky2?

droidicus commented:
@mluessi Did another test on sparky2 with a fresh pull on top of next. Same problem as last time, it logs for a bit and then no more logging after that. Is there anything that I can do to help further debug this?

mluessi commented:
@droidicus I have been testing with sparky2 and can't reproduce what you are seeing; I have been successfully logging at 100Hz with the default profile. I collected several logs to make sure the writing wraps around etc. Are you using a genuine sparky2? Can you post your settings (.uav file) somewhere? Also, what is your test procedure and what peripherals are you using? I have been testing this using sparky2 stand alone using the GCS receiver to arm/disarm.

Unfortunately, sparky2 is pretty bad for logging. It has only 2MB of flash out of which 1MB is used for logging (there are 5 unused 64kB sectors that could also be used, but this needs a BL update). In addition, the [flash chip](http://www.digikey.com/product-detail/en/M25P16-VMN6TP/M25P16-VMN6TPCT-ND/1889049) has a ridiculously long sector erase time of 800ms .. 3s, so unless the file system is freshly formatted, you will get ugly 800ms gaps in the log. This may get better in the future once we add additional buffering, but storing high rate data for 3s will need quite a bit of RAM so I'm not sure if it is realistic.

droidicus commented:
@mluessi, I am 99% sure it is a genuine Sparky2, purchased from http://www.fpvmodel.com/taulabs-sparky2-0-flight-controller_g954.html Good to know that there are some issues with gaps on the Sparky2, this explains some of what I have seen with logging in general. Just to be sure that it wasn't a problem, I did erase the logging partition prior to testing this time.

Here is the UAV dump from my quad: https://gist.github.com/droidicus/3685368785d2ea8841be

My procedure was as follows:
1. Pull/compile/flash the PR
2. Erase the logging partition
3. Boot and enable logging via check box in module widget
4. Power cycle and pre-flight
5. Arm and fly through an autotune
6. Land and disarm
7. Arm and fly for ~10 sec
8. Land, disarm, power cycle
9. Arm and fly for ~10 sec
10. Land and disarm
11. Repeat 6-9 a few more times
12. Connect to GCS, I see the autotune record, and a few of the short flights, but not all of them
13. Disconnect from GCS and power cycle
14. Fly a few more same as 6-9
15. Connect to GCS, no new log records are available (same numbers in the dropdown list on the log download dialog)

Peripherals:
* S.Bus Receiver (FrSky X4R-SB)
* SPort Telemetry

mluessi commented:
@droidicus I was able to reproduce the problem you observed. There was a bug in streamfs that caused it to not properly close files smaller than 1 sector. For sparky2, the sector size is 64kB while for brain it is 4kB, so it was much less likely to be observed on brain. Thanks for all the testing!

droidicus commented:
@mluessi tested again, and it looks good.was able to record multiple logs, and verified that it can wrap after extended logging.

TGTM!

droidicus commented:
Reviewed 1 of 4 files at r3.
Review status: 3 of 7 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/180)*
<!-- Sent from Reviewable.io -->

droidicus commented:
Reviewed 2 of 5 files at r1, 2 of 4 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/180)*
<!-- Sent from Reviewable.io -->

mlyle commented:
Thanks everyone for working to get this landed.  It's gonna be a big win going forward :)
